### PR TITLE
Refactors out common Checksum concept

### PR DESCRIPTION
**Pull request by ryanmoran**
## Summary

We have some common code for splitting and comparing checksums across `cargo` and `postal`. This factors those pieces out into a common place.

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by sophiewigmore** [comment]
I can't comment on it directly for some reason, but there is a linting error in postal/service.go to remove function `determineChecksum`

**Comment by sophiewigmore** [comment]
Also, this seems like a reasonable change but I wanted to point out that introducing the new Checksum type will be breaking for `jam`, since it uses the `cargo.ConfigMetadataDependency` internally

File: cargo/checksum_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		type testCaseType struct {
+			c1     string
+			c2     string
+			result bool
+		}
+
+		for _, tc := range []testCaseType{
+			{"", "", true},
+			{"c", "c", true},
+			{"c", "c", true},
```
Comment (fg-j) [inline cargo/checksum_test.go]: Is this the same test case twice or am I misunderstanding the test setup?

File: cargo/checksum_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			{"sha256:c", "c", true},
+			{"c", "sha256:c", true},
+			{"md5:c", "md5:c", true},
+			{"md5:c", ":c", false},
+			{":", ":", true},
+			{":c", ":c", true},
+			{"", "c", false},
+			{"c", "", false},
+			{"c", "z", false},
+			{"md5:c", "sha256:c", false},
```
Comment (fg-j) [inline cargo/checksum_test.go:35]: For completeness, perhaps it also makes sense to check when the alg is specified the same and the hash is different:
```suggestion
			{"md5:c", "sha256:c", false},
			{"md5:c", "md5:d", false},
```

File: cargo/validated_reader.go [hunk 3]
```diff
@@ -26,28 +25,20 @@ type errorHash struct {
...
-
-	switch splitChecksum[0] {
+	switch checksum.Algorithm() {
 	case "sha256":
 		hash = sha256.New()
 	case "sha512":
 		hash = sha512.New()
 	default:
-		return ValidatedReader{hash: errorHash{err: fmt.Errorf("unsupported algorithm %q: the following algorithms are support [sha256, sha512]", splitChecksum[0])}}
+		return ValidatedReader{hash: errorHash{err: fmt.Errorf("unsupported algorithm %q: the following algorithms are support [sha256, sha512]", checksum.Algorithm())}}
```
Comment (fg-j) [inline cargo/validated_reader.go]:
```suggestion
		return ValidatedReader{hash: errorHash{err: fmt.Errorf("unsupported algorithm %q: the following algorithms are supported [sha256, sha512]", checksum.Algorithm())}}
```

**Review by fg-j — COMMENTED** [review]
A couple nitpicks/clarifications. But overall, LGTM!

**Comment by ryanmoran** [comment]
> I can't comment on it directly for some reason, but there is a linting error in postal/service.go to remove function `determineChecksum`

Good catch. I meant to remove this.

File: cargo/checksum_test.go (hunk 1 above)
Comment (ryanmoran) [reply cargo/checksum_test.go]: Woops. I should have read the existing tests more closely when I refactored this.

**Comment by ryanmoran** [comment]
> Also, this seems like a reasonable change but I wanted to point out that introducing the new Checksum type will be breaking for `jam`, since it uses the `cargo.ConfigMetadataDependency` internally

Ok. I rolled back the changes that might have been breaking. We still use `Checksum` internally and users can cast their checksum string to a `Checksum` to use its helper methods, so I still think this results in a cleaner implementation.

**Review by sophiewigmore — APPROVED** [review]

**Pull request merged**